Allow multiple handlers per queue

zidad opened the pull request:
Fixes behavior that if through naming conventions you get multiple message type handlers per queue, you will not get the exception "no handler registered for this message type".

This is achieved by caching the created HandlerCollection in the factory, and adding a parameter per 

Additional changes:
- Code review fixes based on @micdenny's review for https://github.com/EasyNetQ/EasyNetQ/pull/628/files/ef8238af4be2252ee6da3da30718d3b991b95b4e#diff-f0c2afb9c2755b72eadd3d3a58bb3b2b
- Simplified syntax with new C# features in some places
- Removed region directives
- Fixed the task runner (was throwing errors after .NET core upgrade)

+semver:breaking

micdenny commented:
@zidad I don't see the change for `+semver: breaking` i see the version still `2.0.3` is that correct? probably it is, because until we leave the unstable, probably is correct to leave the version steady, isn't it?

zidad commented:
Hmm there was a space in the comment, I think that might messed it up, let me try to fix that. Can you think of any potential issues with the HandlerCollection caching based on a queue like this, or do you think this should be done in a different way? I could only think of renaming the method to Get... instead of Create... as it's a bit misleading now?

micdenny commented:
> I could only think of renaming the method to Get... instead of Create... as it's a bit misleading now?

sometimes I see name like `GetOrCreate` and this should be sufficient to understand that there's something behind that call and it's not a pure factory or you can simply leave `Create` and hide the intent of the method, in this last way you leave the possibility to completely change mind again, think of a container that uses naming like `Resolve`, you don't know how it will be resolved, you simply know that the service will be returned.

so I would probably avoid to break the contract, because if we will saw performance or unwanted behavior issues, we can revert to a simple straight `new` everycall. 

> Can you think of any potential issues with the HandlerCollection caching based on a queue like this, or do you think this should be done in a different way?

I looked at the code, and this factory is used only when you call `Consume`, so I don't see any big improvements on caching `handlerCollection`, do you?

zidad commented:
> I looked at the code, and this factory is used only when you call Consume, so I don't see any big improvements on caching handlerCollection, do you?

Sorry if this wasn't entirely clear, it's not an optimization, it's a new feature :) 

This has been on my wish list for a while, it allows you to consume multiple message types from the same queue, if you give them the same name through the conventions. See the task runner for an example.

Before this wasn't possible and would throw an exception.

micdenny commented:
> Sorry if this wasn't entirely clear, it's not an optimization, it's a new feature :)

I think I understood the new feature, but I mean this point in the method `public IDisposable Consume(IEnumerable<QueueConsumerPair> queueConsumerPairs, Action<IConsumerConfiguration> configure)`:

```csharp
var handlerCollection = handlerCollectionFactory.CreateHandlerCollection(x.Queue);
```

I think we could avoid the `ConcurrentDictionary` in that factory, because it is not called so many times that point, that change is not difficult to understand nor maintain, but it hides another structure that stays in memory forever to gain quite nothing, am I wrong?

zidad commented:
I'm not sure I'm getting you completely, sorry for explaining. Where would you propose to move the dictionary then? 

We could move it to the RabbitAdvancedBus implementation, I had that first but it looked more messy and it already has way too many responsibilities. You DO need the dictionary somewhere, as this is exactly the thing that fixes the problem (though it doesn't necessarily needs to be a concurrent dictionary in most cases I guess, only if subscribing from different threads?).

Before this dictionary, multiple handler collections would be created for the same queue, meaning your message would get processed correctly only in SOME occasions, but most of the time would fail because it would end up with wrong handler collection.

With the more message types subscribing per queue, the higher the chances were your message would end up in the wrong subscription instance. I'd say it was a bug :)

The dictionary is created only once per bus instance, just like the other services, so I don't see a problem with the long lived structure, or am I missing something?

micdenny commented:
> I'm not sure I'm getting you completely, sorry for explaining wrong. Where would you propose to move the dictionary then?

I didn't understand the need of having the same `HandlerCollection` for same queue, because there's `handlerCollectionFactory.CreateHandlerCollection(x.Queue);` but of course this method can be called more than once and those creating different handlerCollection for the same queue.

That's ok to me, I think that dictionary is in a place where there shouldn't be a lot of load, or at least not so much concurrency, if someone is doing calling `Consume` a lot of time in a concurrent environment, some high contention could be achieved, but I hope is not a so frequent use-case.

zidad commented:
Yeah, I just really prefer the syntax of the ConcurrentDictionary :) It can't do any harm like this right? Are you OK to merge this then?

micdenny commented:
Anyway I'm making some performance test, just in case...

micdenny commented:
@zidad I'm currently investigating an issue about:

> There is already a handler for message type 'PerfMessage'

Same test code, but different behavior, I'm checking what I've done in the perf test...

zidad commented:
Interesting, do you have the sample code published somewhere, and what moment do you get this Exception?

micdenny commented:
With this version I can't subscribe two times on the same type of message.

I think this is quite invasive, not to say an insane drawback 😄 I know that could seems strange to do so, but if I want to create an in-process work queue, I think ENQ should not block you to do it:

```csharp
using System;
using EasyNetQ;

namespace ConsoleApp5
{
    class Program
    {
        static void Main(string[] args)
        {
            var bus = RabbitHutch.CreateBus("host=localhost", x => x.Register<IEasyNetQLogger>(s => new EasyNetQ.Loggers.ConsoleLogger()));

            bus.Subscribe<MyMessage>("Test", message =>
            {
            });

            bus.Subscribe<MyMessage>("Test", message =>
            {
            });

            Console.ReadLine();

            bus.Dispose();
        }
    }

    public class MyMessage { }
}
```

This code is completely legit, and it means that RMQ will load balancing the messages to that two subscription in the same process.

micdenny commented:
BTW the performance seems to be good, but about the blocking on multiple subscriber, I'm not sure at all.

zidad commented:
Gotcha, nice catch. Then I'll just add the factory as a separate implementation then and leave the original as default. Even though I think the possibility of having multi-purpose (less) queues might be a more likely use case (it's more like the MassTransit 2.x behavior of having a single queue per process for all messages by default).

What's the value of in-process load balancing, if it's all a single threaded queue anyway, or does this have some kind of prioritization purpose?

>  but about the blocking on multiple subscriber, I'm not sure at all.

What exactly do you mean by that? that it doesn't necessarily need a concurrent dictionary? Can't do harm either right?

micdenny commented:
I drop also this small typo I found:

private readonly ICollection<Tuple<IQueue, Func<byte[], MessageProperties, MessageReceivedInfo, Task>>> _queueConsumerPairs;

the name of private members must be without `_`, I like the `_` way, but this code-base was without, so it's better to follow that naming convention.

micdenny commented:
> it's more like the MassTransit 2.x behavior of having a single queue per process for all messages by default

I never intensively played with MassTransit (my fault), but I think we are not ready to change direction at this stage, at least not in this way, I mean, with this PR we want to introduce a new feature, not change another one, I know we are creating the v2, but I think this is something for another breaking change and should be for a >=v3 edition, because there's a lot of decision to take if we want to follow another direction (maybe better) like mass transit v2 has taken, but I think we need more time, at least personally, to play around mass transit and better understand what we are talking about.

> What's the value of in-process load balancing, if it's all a single threaded queue anyway, or does this have some kind of prioritization purpose?

Personally I never did it, but I think if we want to change this behavior, we should think twice, so it means at least separate it from this feature PR, and open another issue that describes we want to cut that feature/behavior to simplify the code-base and open other possibilities, you can open it now with an associated PR, and then re-discuss it there and in the meanwhile to **close this feature doing as you said, adding one more method to preserve what is the behavior right now** (even if this will break the ones that are implementing that interface, I don't think so much people 😄 ).

Another strange behavior I'm seeing is this: even when `Dispose()` is called on the subscription, I get the exception trying to subscribe again, do you think this should be accepted? In my opinion yes, because otherwise you fall in a limbo, where you think you have disposed a resource and you can re-issue the command, but you can't:

```csharp
using System;
using EasyNetQ;

namespace ConsoleApp5
{
    class Program
    {
        static void Main(string[] args)
        {
            var bus = RabbitHutch.CreateBus("host=localhost", x => x.Register<IEasyNetQLogger>(s => new EasyNetQ.Loggers.ConsoleLogger()));

            var subscription = bus.Subscribe<MyMessage>("Test", message =>
            {
            });

            subscription.Dispose();

            bus.Subscribe<MyMessage>("Test", message =>
            {
            });

            Console.ReadLine();

            bus.Dispose();
        }
    }

    public class MyMessage { }
}
```

see the `subscription.Dispose();` in the middle.


> What exactly do you mean by that? that it doesn't necessarily need a concurrent dictionary? Can't do harm either right?

I meant the fact that you can't subscribe two times, I was simply repeating myself, forget about this sentence, I'm often quite verbose 😄

zidad commented:
Updated, added the factory as a separate class. 

I've also removed the underscore naming prefix in most other places

I haven't found an easy way to remove the handler from the factory though if it gets disposed, maybe it's possible through the event bus? 

Not sure if it's needed right now as this is no longer default behavior? 

I'd really like to use this behavior as-is so I'm eager to merge it :)

micdenny commented:
> I haven't found an easy way to remove the handler from the factory though if it gets disposed, maybe it's possible through the event bus?
> Not sure if it's needed right now as this is no longer default behavior?

we can postpone it because we have a new method overload for this new feature and the old method remains as is, but anyway I think the dispose is a feature that is used in that way, I have a real scenario, where when a user connect to a real-time api based on signalR I subscribe to its queue, when it disconnect I dispose the subscription so its messages stays on its queue until it connects again.

> I'd really like to use this behavior as-is so I'm eager to merge it :)

I run tests again and let you know...

micdenny commented:
@zidad performance test are ok, you can merge it

zidad commented:
Hmm that semver comment actually increased the major version. Trying to reverse it, do you have the possibility to delete a package from nuget?

micdenny commented:
No, only to unlist :(

zidad commented:
That should be fine too, would you mind unlisting the 3.x versions?

micdenny commented:
The only problem I see is that we cannot re-use that specific version, using the build number as we do will probably avoid the problem 

I'm going to unlist it

zidad commented:
Thanks!

micdenny commented:
ok i've unlisted `3.0.0-unstable0059`

zidad commented:
Great... Could you do the same for the other packages? (EasyNetQ.DI.* and EasyNetQ.Serilog)... Sorry for the mess, I thought breaking would update the smaller version number 👎

micdenny commented:
> Could you do the same for the other packages?

done!

> I thought breaking would update the smaller version number

In semver a breaking change must always increment the major, before in ENQ was different.

http://semver.org/

> Given a version number MAJOR.MINOR.PATCH, increment the:
> 
> MAJOR version when you make incompatible API changes,
> MINOR version when you add functionality in a backwards-compatible manner, and
> PATCH version when you make backwards-compatible bug fixes.
